app/utils/service_client/middleware/middleware_base.rb

```diff
@@ -0,0 +1,36 @@
+module ServiceClient
+  module Middleware
+
+    # Base class for middlewares.
+    #
+    # Default implementations for `enter`, `leave` and `error`
+    # return the context object `ctx` without manipulating it.
+    # The subclasses can override all or some of these methods.
```

Comment:
Before invoking a lifecycle method like enter, leave or error the context runner implementation checks if a middleware implements it. At each stage, only middleware that respond to the lifecycle method in question is invoked. Inheriting from this class breaks the idea because now everybody implement every lifecycle method. I would suggest making this a marker class only and either not inherit from it or then leave out these default implementations of leave, enter and error.